pallets/pool-fees/src/lib.rs

```diff
@@ -0,0 +1,810 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub mod types;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cfg_traits::{
+		changes::ChangeGuard,
+		fee::{AddPoolFees, FeeAmountProration},
+		EpochTransitionHook, PoolInspect, PoolNAV, PoolReserve, PreConditions, Seconds, TimeAsSecs,
+	};
+	use cfg_types::{
+		pools::{
+			PoolFee, PoolFeeAmount, PoolFeeAmounts, PoolFeeBucket, PoolFeeEditor, PoolFeeInfo,
+			PoolFeeType,
+		},
+		portfolio,
+		portfolio::{InitialPortfolioValuation, PortfolioValuationUpdateType},
+	};
+	use frame_support::{
+		pallet_prelude::*,
+		traits::{
+			fungibles::{Inspect, Mutate},
+			tokens,
+		},
+		weights::Weight,
+		PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+	use parity_scale_codec::HasCompact;
+	use sp_arithmetic::{
+		traits::{EnsureAdd, EnsureAddAssign, EnsureSubAssign, One, Saturating, Zero},
+		ArithmeticError, FixedPointOperand,
+	};
+	use sp_runtime::{traits::AccountIdConversion, SaturatedConversion};
+	use sp_std::vec::Vec;
+	use strum::IntoEnumIterator;
+
+	use super::*;
+	use crate::types::Change;
+
+	pub type PoolFeeInfoOf<T> = PoolFeeInfo<
+		<T as frame_system::Config>::AccountId,
+		<T as Config>::Balance,
+		<T as Config>::Rate,
+	>;
+
+	pub type PoolFeeOf<T> = PoolFee<
+		<T as frame_system::Config>::AccountId,
+		<T as Config>::FeeId,
+		PoolFeeAmounts<<T as Config>::Balance, <T as Config>::Rate>,
+	>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The identifier of a particular fee
+		type FeeId: Parameter
+			+ Member
+			+ Default
+			+ TypeInfo
+			+ MaxEncodedLen
+			+ Copy
+			+ EnsureAdd
+			+ One
+			+ Ord;
+
+		/// The source of truth for the balance of accounts
+		type Balance: tokens::Balance + FixedPointOperand + From<Seconds>;
```

Comment:
NIT: Maybe instead of being converted from `Seconds`, you could want to say that balance is divisible by `Seconds`? Something like: `Div<Seconds>` or similar?

Reply:
AFAICS, we require `Balance` to at least impl `Mul<Seconds>` and `Div<Seconds>`. Given that `Seconds` is `u64` and `Balance` always `u128`, I think this is fine.